backend/composer/services/graph_service.py

```diff
@@ -0,0 +1,128 @@
+JOURNEY_DELIMITER = '\\'
+
+
+def generate_paths(origins, vias, destinations):
+    paths = []
+    # Calculate the total number of layers, including origins and destinations
+    number_of_layers = len(vias) + 2 if vias else 2
+
+    # Handle direct connections from origins to destinations
+    for origin in origins:
```

Comment:
@afonsobspinto this is very "3rd generation" development
can't we use the database to retrieve this and remove the for loops ?

Reply:
Without changing the model to contain information about which **layer** we are connecting to (or from) I don't think we can get rid of these for loops unless we replace them with custom sql queries, but I'm not sure that's what you are aiming for here. Would you mind to elaborate a bit on what's your idea @zsinnema ?

Reply:
The core problem / complexity of this service is that we don't have an exact graph information. 
We know that a node is connected to some entity from a previous layer but not exactly to each one. 

Roughly, in order to beat this problem, I'm going layer by layer and constructing the paths as we go.

Reply:
The second part of the algorithm takes all those paths found and 'consolidates' them.

Reply:
@afonsobspinto currently I have no hours left to work on this. My suggestion would be to retrieve all "data" at once and then use python to structure it. Please check the debug toolbar queries when retrieving the connectivity statements. It will show you that this request is executing hundreds/thousands queries. The goal here would be to reduce the number of queries for retrieving CS's

Reply:
I understand [7ce58de](https://github.com/MetaCell/sckan-composer/pull/189/commits/7ce58de75f731129c70221e84e46440e4b9204c6) should solve the many database queries.